posts/wit-bringing-types-to-near-contracts.md

```diff
@@ -0,0 +1,310 @@
+  return contract.get_message();
+}
+
+
+```
+
+Since the original comments in the rust code are preserved you can also get hover over docs in your IDE, or generate the docs see [TenK's docs](https://tenk-dao.github.io/tenk/docs/).
+
+## JSON Schema
+
+Taking this a step further we can generate a [json-schema](https://json-schema.org/), basically a JSON object that defines the constraints of the data to allow a JSON object to be validated.
+
+```bash
+witme near json
+```
+
+Currently this is supported by using a [ts-json-schema-generator](https://github.com/vega/ts-json-schema-generator). Though this too could be generated directly from the `.wit`.  This command defaults to find a `./ts/index.ts`, which it uses to generate a `index.schema.json`.
+
+Which would look something like
+
+```json
+{
+  "GetMessage": {
+    "additionalProperties": false,
+    "contractMethod": "view",
+    "description": "A view call to get the current message",
+    "properties": {
+      "args": {
+        "additionalProperties": false,
+        "type": "object"
+      }
+    },
+    "required": [
+      "args"
+    ],
+    "type": "object"
+  },
+ "SetMessage": {
+      "additionalProperties": false,
+      "contractMethod": "change",
+      "description": "A change call to set the message",
+      "properties": {
+        "args": {
+          "additionalProperties": false,
+          "properties": {
+            "message": {
+              "$ref": "#/definitions/Message"
+            }
+          },
+          "required": [
+            "message"
+          ],
+          "type": "object"
+        },
+        "options": {
+          "additionalProperties": false,
+          "properties": {
+            "attachedDeposit": {
+              "$ref": "#/definitions/Balance",
+              "default": "0",
+              "description": "Units in yoctoNear"
+            },
+            "gas": {
+              "default": "30000000000000",
+              "description": "Units in gas",
+              "pattern": "[0-9]+",
+              "type": "string"
+            }
+          },
+          "type": "object"
+        }
+      },
+      "required": [
+        "args",
+        "options"
+      ],
+      "type": "object"
+    },
+  "Message": {
+      "additionalProperties": false,
+      "description": "A message that contains some text",
+      "properties": {
+        "text": {
+          "description": "Inner string value",
+          "type": "string"
+        }
+      },
+      "required": [
+        "text"
+      ],
+      "type": "object"
+    },
+}
+```
+
+First we can see `GetMessage`, is a `view` function and requires an `args` object that has no properties. Next `SetMessage` requires an `args` object with one field "`message`", the type of which is a reference to the `Message` type defined in the schema. Since `SetMeassage` is a `change` function an `options` field is also required for how much gas and deposit to attached to the transaction.
```

Comment:
```suggestion
First we can see `GetMessage` is a `view` function and requires an `args` object that has no properties. Next `SetMessage` requires an `args` object with one field "`message`", the type of which is a reference to the `Message` type defined in the schema. Since `SetMessage` is a `change` function, an `options` field is also required for how much gas and deposit to attached to the transaction.
```